src/clients/Deluge.ts

```diff
@@ -0,0 +1,232 @@
+		});
+
+		if (response === null) {
+			throw new CrossSeedError(
+				`failed to establish a connection to deluge: ${this.delugeWebUrl.origin}`
+			);
+		}
+	}
+
+	/**
+	 * ensures authentication and sends JSON-RPC calls to deluge
+	 */
+	private async call(body: { method: string; params: any[] }) {
+		this.msgId = (this.msgId + 1) % 1024;
+
+		const headers = new Headers({ "Content-Type": "application/json" });
+		if (this.delugeCookie) headers.set("Cookie", this.delugeCookie);
+
+		const url = this.delugeWebUrl.origin + this.delugeWebUrl.pathname;
+		let response: Response, json: DelugeResponse;
+		try {
+			response = await fetch(url, {
+				body: JSON.stringify({ ...body, id: this.msgId }),
+				method: "POST",
+				headers: headers,
+			});
+			json = (await response.json()) as DelugeResponse;
+		} catch (networkError) {
+			return null;
+		}
+
+		if (json && json.error && json.error.code && json.error.code === 1) {
+			await this.authenticate();
+			try {
+				response = await fetch(url, {
+					body: JSON.stringify({ ...body, id: this.msgId }),
+					method: "POST",
+					headers: headers,
+				});
+			} catch (networkError) {
+				return null;
+			}
+			json = (await response.json()) as DelugeResponse;
+		}
+		if (!response.ok) {
+			return null;
+		}
+
+		if (response.headers && response.headers.has("Set-Cookie")) {
+			this.delugeCookie = response.headers
+				.get("Set-Cookie")
+				.split(";")[0];
+		} else if (body.method == "auth.login") {
+			throw new CrossSeedError(
+				`failed to authenticate the session in deluge: ${this.delugeWebUrl.origin}`
+			);
+		}
+		return json;
+	}
+
+	/**
+	 * checks enabled plugins for "Label"
+	 * returns true if successful.
+	 */
+	private async labelEnabled() {
+		const enabledLabels = await this.call({
+			method: "core.get_enabled_plugins",
+			params: [],
+		});
+		return enabledLabels.result && enabledLabels.result.includes("Label");
+	}
+
+	/**
+	 * if Label plugin is loaded, adds (if necessary)
+	 * and sets the label based on torrent hash.
+	 */
+	private async setLabel(infoHash: string): Promise<void> {
+		if (this.isLabelEnabled) {
+			const setResult = await this.call({
+				method: "label.set_torrent",
+				params: [infoHash, this.delugeLabel],
+			});
+			if (
+				setResult.error &&
+				setResult.error.code &&
+				setResult.error.code == 4
+			) {
+				await this.call({
+					method: "label.add",
+					params: [this.delugeLabel],
+				});
+				await this.call({
+					method: "label.set_torrent",
+					params: [infoHash, this.delugeLabel],
+				});
+			}
+		}
+	}
+
+	/**
+	 * injects a torrent into deluge client
+	 */
+	async inject(
+		newTorrent: Metafile,
+		searchee: Searchee,
+		path?: string
+	): Promise<InjectionResult> {
+		if (searchee.infoHash && !(await this.checkCompleted(searchee))) {
+			return InjectionResult.TORRENT_NOT_COMPLETE;
+		}
+		const params = this.formatData(
+			`${newTorrent.name}.cross-seed.torrent`,
+			newTorrent.encode().toString("base64"),
+			path
```

Comment:
unfortunately, `path` isn't guaranteed to exist (for torrent based matching, we query the client for `searchee.infoHash` to get its download location and mirror that on the new torrent